tokio-tls/src/lib.rs

```diff
@@ -33,76 +46,217 @@ use tokio_io::{try_nb, AsyncRead, AsyncWrite};
-pub struct Accept<S> {
-    inner: MidHandshake<S>,
+// pub struct Accept<S> {
+//     inner: MidHandshake<S>,
+// }
+
+struct MidHandshake<S>(Option<MidHandshakeTlsStream<AllowStd<S>>>);
+
+enum StartedHandshake<S> {
+    Done(TlsStream<S>),
+    Mid(MidHandshakeTlsStream<AllowStd<S>>),
+}
+
+struct StartedHandshakeFuture<F, S>(Option<StartedHandshakeFutureInner<F, S>>);
+struct StartedHandshakeFutureInner<F, S> {
+    f: F,
+    stream: S,
+}
+
+impl<S> AllowStd<S>
+where
+    S: Unpin,
+{
+    fn with_context<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut Context<'_>, Pin<&mut S>) -> R,
+    {
+        unsafe {
+            assert_ne!(self.context, 0);
+            let waker = &mut *(self.context as *mut _);
+            f(waker, Pin::new(&mut self.inner))
+        }
+    }
 }
 
-struct MidHandshake<S> {
-    inner: Option<Result<native_tls::TlsStream<S>, HandshakeError<S>>>,
+impl<S> Read for AllowStd<S>
+where
+    S: AsyncRead + Unpin,
+{
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self.with_context(|ctx, stream| stream.poll_read(ctx, buf)) {
+            Poll::Ready(r) => r,
+            Poll::Pending => Err(io::Error::from(io::ErrorKind::WouldBlock)),
+        }
+    }
+}
+
+impl<S> Write for AllowStd<S>
+where
+    S: AsyncWrite + Unpin,
+{
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self.with_context(|ctx, stream| stream.poll_write(ctx, buf)) {
+            Poll::Ready(r) => r,
+            Poll::Pending => Err(io::Error::from(io::ErrorKind::WouldBlock)),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        match self.with_context(|ctx, stream| stream.poll_flush(ctx)) {
+            Poll::Ready(r) => r,
+            Poll::Pending => Err(io::Error::from(io::ErrorKind::WouldBlock)),
+        }
+    }
+}
+
+fn cvt<T>(r: io::Result<T>) -> Poll<io::Result<T>> {
+    match r {
+        Ok(v) => Poll::Ready(Ok(v)),
+        Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => Poll::Pending,
+        Err(e) => Poll::Ready(Err(e)),
+    }
 }
 
 impl<S> TlsStream<S> {
     /// Get access to the internal `native_tls::TlsStream` stream which also
     /// transitively allows access to `S`.
-    pub fn get_ref(&self) -> &native_tls::TlsStream<S> {
-        &self.inner
+    pub fn get_ref(&self) -> &native_tls::TlsStream<AllowStd<S>> {
+        &self.0
     }
 
     /// Get mutable access to the internal `native_tls::TlsStream` stream which
     /// also transitively allows mutable access to `S`.
-    pub fn get_mut(&mut self) -> &mut native_tls::TlsStream<S> {
-        &mut self.inner
+    pub fn get_mut(&mut self) -> &mut native_tls::TlsStream<AllowStd<S>> {
+        &mut self.0
+    }
+
+    fn with_context<F, R>(&mut self, ctx: &mut Context<'_>, f: F) -> R
+    where
+        F: FnOnce(&mut native_tls::TlsStream<AllowStd<S>>) -> R,
+        AllowStd<S>: Read + Write,
+    {
+        self.0.get_mut().context = ctx as *mut _ as usize;
```

Comment:
This function is dangerous in that if `f` panics, then `TlsStream` is left in an unsound state.

The strategy is to use a guard that resets `context` to `0` on drop. Here is an example of one: https://github.com/tokio-rs/tokio/blob/master/tokio-executor/src/global.rs#L160-L168

It might end up being a bit more complicated in this case to get the lifetimes to work... you could prob ably do:

```rust
struct Guard<'a, S>(&'a mut TlsStream);

impl<'a, S> Drop for Guard<'a, S> {
    fn drop(&mut self) {
        self.0.0.get_mut().context = 0;
    }
}

self.0.get_mut().context = ctx as *mut _ as usize;
let g = Guard(self);
f(&mut g.0.0);
```